theano/gpuarray/c_code/dnn_gi.c

```diff
@@ -159,11 +185,12 @@ APPLY_SPECIFIC(conv_gi)(PyGpuArrayObject *kerns, PyGpuArrayObject *output,
 
   std::string hashkey;
 
-  size_t maxfree = c_get_largest_free_block_size(c);
-  if (PyErr_Occurred()) return 1;
 
   cuda_enter(c->ctx);
 
+  size_t maxfree = c_get_largest_free_block_size(c);
+  if (PyErr_Occurred()) return 1;
```

Comment:
We would need to exit the context. I think this is the reason why it was before cuda_enter()

Reply:
Indeed. I changed the place to have same code layout as in `dnn_fwd`, but even in `dnn_fwd` we forget to exit the context.

I will exit context properly in all 3 files.